ports/esp32/machine_can.c

```diff
@@ -0,0 +1,673 @@
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+#include <string.h>
+
+#include "py/obj.h"
+#include "py/objarray.h"
+#include "py/binary.h"
+#include "py/runtime.h"
+#include "py/builtin.h"
+#include "py/mphal.h"
+#include "py/mperrno.h"
+#include "mpconfigport.h"
+
+// Headers of ESP-IDF library
+#include "soc/dport_reg.h"
+#include "driver/can.h"
+#include "esp_err.h"
+#include "esp_log.h"
+
+#include "modmachine.h"
+
+#if MICROPY_HW_ENABLE_CAN
+
+#define DEVICE_NAME "CAN"
+
+#define CAN_BAUDRATE_25k 25
+#define CAN_BAUDRATE_50k 50
+#define CAN_BAUDRATE_100k 100
+#define CAN_BAUDRATE_125k 125
+#define CAN_BAUDRATE_250k 250
+#define CAN_BAUDRATE_500k 500
+#define CAN_BAUDRATE_800k 800
+#define CAN_BAUDRATE_1M 1000
+
+#define ESP_STATUS_CHECK(status)   \
+    if (status != ESP_OK)          \
+    {                              \
+        mp_raise_OSError(-status); \
+    }
+typedef enum _filter_mode_t{
+    FILTER_RAW_SINGLE = 0,
+    FILTER_RAW_DUAL,
+    FILTER_ADDRESS
+} filter_mode_t;
+
+typedef struct _machine_can_config_t{
+    can_timing_config_t timing;
+    can_filter_config_t filter;
+    can_general_config_t general;
+    uint16_t baudrate;
+    bool initialized;
+} machine_can_config_t;
+
+typedef struct _machine_can_obj_t{
+    mp_obj_base_t base;
+    machine_can_config_t *config;
+    mp_obj_t rxcallback;
+    byte rx_state;
+    bool extframe : 1;
+    bool loopback : 1;
+    uint16_t num_error_warning; //FIXME: populate this value somewhere
+    uint16_t num_error_passive;
+    uint16_t num_bus_off;
+} machine_can_obj_t;
+
+typedef enum _rx_state_t{
+    RX_STATE_FIFO_EMPTY = 0,
+    RX_STATE_MESSAGE_PENDING,
+    RX_STATE_FIFO_FULL,
+    RX_STATE_FIFO_OVERFLOW,
+} rx_state_t;
+
+// Default baudrate: 500kb
+#define CAN_DEFAULT_PRESCALER (8)
+#define CAN_DEFAULT_SJW (3)
+#define CAN_DEFAULT_BS1 (15)
+#define CAN_DEFAULT_BS2 (4)
+
+// Internal Functions
+mp_obj_t machine_hw_can_make_new(const mp_obj_type_t *type, size_t n_args, size_t n_kw, const mp_obj_t *args);
+STATIC mp_obj_t machine_hw_can_init_helper(machine_can_obj_t *self, size_t n_args, const mp_obj_t *pos_args, mp_map_t *kw_args);
+STATIC void machine_hw_can_print(const mp_print_t *print, mp_obj_t self_in, mp_print_kind_t kind);
+
+// singleton CAN device object
+machine_can_config_t can_config = {.general = CAN_GENERAL_CONFIG_DEFAULT(2, 4, 0),
+                                   .filter = CAN_FILTER_CONFIG_ACCEPT_ALL(),
+                                   .timing = CAN_TIMING_CONFIG_25KBITS(),
+                                   .initialized = false};
+
+STATIC machine_can_obj_t machine_can_obj = {{&machine_can_type}, .config = &can_config};
+
+// INTERNAL FUNCTION Return status information
+STATIC can_status_info_t _machine_hw_can_get_status(){
```

Comment:
There isn't really a convention for leading-underscore naming for internal stuff. And the comment probably isn't needed -- making it static is a clear marker of what's going on.